### PR TITLE
Update graph construction design doc

### PR DESCRIPTION
**Pull request by wangkuiyi**
Fixes https://github.com/PaddlePaddle/Paddle/issues/3861

### CODE REVIEW

File: doc/design/graph.md [hunk 1]
```diff
@@ -25,7 +27,9 @@ The first four lines of above program build the forward part of the graph.
 
 ![](images/graph_construction_example_forward_only.png)
 
-In particular, the first line `x = layer.data("images")` creates variable x and a Feed operator that copies a column from the minibatch to x.  `y = layer.fc(x)` creates not only the FC operator and output variable y, but also two parameters, W and b.
+In particular, the first line `x = layer.data("images")` creates variable x and a Feed operator that copies a column from the minibatch to x.  `y = layer.fc(x)` creates not only the FC operator and output variable y, but also two parameters, W and b, and the initialization operators.
+
+Initialization operators are kind of "run-once" operators -- the `Run` method increments a class data member counter so to run at most once.  By doing so, a parameter wouldn't be initialized repeatedly, say, in every minibatch.
```
Comment (jacquesqiao) [inline doc/design/graph.md]: run-once maybe not a very good choice. Because sometimes the user may want to reinitialize the params. Maybe we should think out some better way to do it.

File: doc/design/graph.md [hunk 2]
```diff
@@ -49,3 +53,18 @@ According to the chain rule of gradient computation, `ConstructBackwardGraph` wo
 For each parameter, like W and b created by `layer.fc`, marked as double circles in above graphs, `ConstructOptimizationGraph` creates an optimization operator to apply its gradient.  Here results in the complete graph:
 
 ![](images/graph_construction_example_all.png)
```
Comment (helinwang) [inline doc/design/graph.md]: Currently we only have **one** kind of directed edge. Which represents both dependency and data flow. It works well but on some case it have problem:
I think we should call A depends on B only if in **every** step running A requires running B.
Not every step running "FC" requires "init" (only the first step needs init),  But currently according to the graph "FC" depends on "W", and "W" depends on "init", thus "FC" depends on "init". I think this kind of "before-after" relationship should be specified by the user, not the graph.

In my opinion we need **two** kinds of directed edges. One for dependency, one for data flow. And maybe for discussion we don't need to draw intermediate var. In the below graph the dotted line is data flow, the full line is dependency. In this representation, there is no cycle in the graph, and "FC" no longer depends on "init" (so we don't need to come up a way to let "init" only run once).

screen shot 2017-09-04 at 4 45 57 pm

Comment (wangkuiyi) [reply doc/design/graph.md]: Got it. I love this idea and the figure! I agree that there are two kinds of dependencies -- the data dependency and the execution dependency. Currently, we treat them as the same and represent them by the order of operators in array `repeated OpDesc ops` in protobuf message BlockDesc.

I am not sure if it is necessary to explicitly describe these two kinds of dependencies in our protobuf messages.  A reason is that I am not sure what InitAll is -- is it a Var like those returned by operator binding functions, or is it an operator?

File: doc/design/graph.md (hunk 1 above)
Comment (wangkuiyi) [reply doc/design/graph.md]: Agreed. It would be great if we can have another solution.  How about we keep the run-once operator as a viable solution right now, and update it later after we got a better idea?

File: doc/design/graph.md [hunk 3]
```diff
@@ -8,6 +8,8 @@ This document explains that the construction of a graph as three steps:
 - construct the backward part
 - construct the optimization part
 
+## The Construction of a Graph
```
[the remark this answers is not in the document]
Comment (wangkuiyi) [reply doc/design/graph.md]: I think the training needs 1) the cost, and 2) the parameter to be optimized to minimize the cost.

The cost is specified in the invocation to `train`.

Parameters could be created by a layer function like `layer.fc`, or the user via `W = paddle.Var(type=parameter, ...)`. Anyway, they are marked parameters and can be updated.

File: doc/design/graph.md (hunk 2 above)
Comment (helinwang) [reply doc/design/graph.md]: Sorry, I should have make "init all" more clear. It's an OP that joins / merges all the dependency: will run when all its dependencies are done, it's does nothing itself (only used to join the dependency). Maybe we can call it join or merge.

The reason behind why we need to explicitly describe these two kinds of dependencies is: the PaddlePaddle scheduler only need to schedule OP to run according to the dependency constraint (rather than data flow). For example, in this case, even though "init" writes to var "B", var "B" no longer depends on "init", so when doing optimization, "init" will not be scheduled.

File: doc/design/graph.md (hunk 1 above)
Comment (helinwang) [reply doc/design/graph.md]: https://github.com/PaddlePaddle/Paddle/pull/3862#discussion_r136876843 could solve it :)

File: doc/design/graph.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (helinwang) [reply doc/design/graph.md]: > the training needs 1) the cost, and 2) the parameter to be optimized to minimize the cost.

I think it need the optimizer as well (Adam or Adagrad).
For example, it user do something like:
```Python
opt0 = pd.Adam(cost)
opt1 = pd.Adagrad(cost)
train(cost, reader=mnist.train())
```
What optimizer will Paddle use for training?
However, maybe the Python code you wrote is the V2 API, which maybe only allow one optimizer to be connected with the cost.

File: doc/design/graph.md (hunk 2 above)
Comment (helinwang) [reply doc/design/graph.md]: Another solution is TF's solution: there is no type Var in the graph. A graph only has OP. A var is represented by a "var OP", which only have output (output the handle for read / write), but no input. And every edge is a tensor rather than var:
screen shot 2017-09-05 at 12 47 42 pm

Comment (wangkuiyi) [reply doc/design/graph.md]: Record the temporary conclusions from offline discussions:

1. TensorFlow's graph representation embeds variables into operators, and
1. requires users specify input, output, and dependent operators for each operator.

The specification of dependencies looks ugly.  So let's follow our current design of using variables and operators.

File: doc/design/graph.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (wangkuiyi) [reply doc/design/graph.md]: Yes. What I mean is that we can have two forms of `Block::Eval`:

1. One accepts targets of type `Variable`s: 

   ```c++
   void Block::Eval(vector<Variable*> targets);
   ```

   which is used to do forward computation.  It traces only operators in `BlockDesc::ops` before `targets`.

    1. *Forward computation*: Because our Python API doesn't expose gradient variables to users, `targets` have to be forward variables, so this form of `Block::Eval` works only with forward computation.

   1. *Backward computation*: In the C++ world, `Block::Eval` can accept gradient variables as its targets. We can create a Python API function, say `backward`, which calls `Block::Eval` with gradient variables to do the backward computation.

1. The other form of `Block::Eval` accepts targets as operators:

   ```c++
   void Block::Eval(vector<Operator*> targets);
   ```

   Somewhere in the C++ world, we can enumerate all optimization operators and use them as the target, so could we run the optimization step.
